File: cpp/medium/0380_insert_delete_getrandom_o1.cpp
// https://leetcode.com/problems/insert-delete-getrandom-o1/description/?envType=study-plan-v2&envId=top-interview-150
class RandomizedSet {
    private:
        vector<int> set;
        unordered_map<int, int> map;
public:
    RandomizedSet() {
    }
    
    bool insert(int val) {
        if(map.count(val) > 0){
            return false;
        } else{
            map[val] = set.size();
            set.push_back(val);
            return true;
        }
    }
    
    bool remove(int val) {
        if(map.count(val) == 0){
            return false;
        } else{
            int lastVal = set[set.size() - 1];
            map[lastVal] = map[val];
            set[map[val]] = lastVal;
            set.pop_back();
            map.erase(val);
            return true;
        }
    }
    
    int getRandom() {
        return set[rand() % set.size()];
    }
};

File: python/medium/0380_insert_delete_getrandom_o1.py
# https://leetcode.com/problems/insert-delete-getrandom-o1/description/
class RandomizedSet:

    def __init__(self):
        self.valSet = []
        self.valMap = {}

    def insert(self, val: int) -> bool:
        if val in self.valMap:
            return False
        else:
            self.valSet.append(val)
            self.valMap[val] = len(self.valSet) - 1
            return True

    def remove(self, val: int) -> bool:
        if val in self.valMap:
            index = self.valMap[val]

            lastKey = self.valSet[-1]
            self.valSet[index] = lastKey
            self.valMap[lastKey] = index

            self.valSet.pop()
            self.valMap.pop(val)
            return True
        else:
            return False


    def getRandom(self) -> int:
        return random.choice(self.valSet)
